pkg/auth/token_handler.go

```diff
@@ -0,0 +1,259 @@
+	featureFlags := c.featureStore.Load()
+	if !featureFlags.IsOIDCAuthentication() {
+		// we can skip if OIDC is not enabled
+		c.logger.Debug("oidc authentication is disabled, so skipping to update provider config")
+		return nil
+	}
+
+	c.logger.Debug("updating OIDC provider config...")
+
+	ctx := context.Background()
+
+	// load OIDC configuration from API server
+	oidcConfig, err := c.loadKubernetesOIDCConfiguration(ctx)
+	if err != nil {
+		return fmt.Errorf("could not get clusters OIDC configuration: %w", err)
+	}
+
+	if strings.HasPrefix(oidcConfig.Issuer, "https://kubernetes.default.svc") {
+		// setup http client for OIDC lib with the CA of the API server
+		httpClient, err := c.getHTTPClientForKubeAPIServer()
+		if err != nil {
+			return fmt.Errorf("could not get HTTP client with TLS certs of API server: %w", err)
+		}
+		ctx = oidc.ClientContext(context.Background(), httpClient)
+	} else {
+		// TODO: if OIDC issuer is an external one, configure http client with certs for external issuer
+		// (e.g. load certs from a config map)
+	}
+
+	// get OIDC provider
+	provider, err := oidc.NewProvider(ctx, oidcConfig.Issuer)
+	if err != nil {
+		return fmt.Errorf("could not get OIDC provider: %w", err)
+	}
+
+	c.providerLock.Lock()
+	defer c.providerLock.Unlock()
+
+	c.provider = provider
+
+	c.logger.Debug("updated OIDC provider config", zap.Any("oidc-config", oidcConfig))
+
+	return nil
+}
+
+func (c *OIDCTokenHandler) getHTTPClientForKubeAPIServer() (*http.Client, error) {
+	cfg, err := rest.InClusterConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not get in cluster config: %w", err)
+	}
+
+	client, err := rest.HTTPClientFor(cfg)
+	if err != nil {
+		return nil, fmt.Errorf("could not create HTTP client from rest config: %w", err)
+	}
+
+	return client, nil
+}
+
+func (c *OIDCTokenHandler) loadKubernetesOIDCConfiguration(ctx context.Context) (*openIDMetadata, error) {
+	req := c.kubeClient.Discovery().RESTClient().Get().RequestURI("/.well-known/openid-configuration")
+
+	res := req.Do(ctx)
+	if err := res.Error(); err != nil {
+		return nil, fmt.Errorf("could not execute request: %w", err)
+	}
+
+	body, err := res.Raw()
+	if err != nil {
+		return nil, fmt.Errorf("could not get response: %w", err)
+	}
+
+	openIdConfig := &openIDMetadata{}
+	if err := json.Unmarshal(body, openIdConfig); err != nil {
+		return nil, fmt.Errorf("could not unmarshall openid config: %w", err)
+	}
+
+	return openIdConfig, nil
+}
+
+type IDToken struct {
+	Issuer          string
+	Audience        []string
+	Subject         string
+	Expiry          time.Time
+	IssuedAt        time.Time
+	AccessTokenHash string
+}
+
+type openIDMetadata struct {
+	Issuer        string   `json:"issuer"`
+	JWKSURI       string   `json:"jwks_uri"`
+	ResponseTypes []string `json:"response_types_supported"`
+	SubjectTypes  []string `json:"subject_types_supported"`
+	SigningAlgs   []string `json:"id_token_signing_alg_values_supported"`
+}
```

Comment:
A similar logic is in `NewProvider` from go-oidc, can't we reuse more of that?

Reply:
Unfortunately the structs are private. And we need to load the oidc discovery config to set the issuer, if it's not `https://kubernetes.default.svc` (external issuer case).